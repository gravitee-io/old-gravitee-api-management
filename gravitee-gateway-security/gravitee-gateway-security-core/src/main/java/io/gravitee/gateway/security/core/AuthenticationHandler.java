/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.gateway.security.core;

import io.gravitee.gateway.api.ExecutionContext;

import java.util.List;

/**
 * @author David BRASSELY (david.brassely at graviteesource.com)
 * @author GraviteeSource Team
 */
public interface AuthenticationHandler {

    /**
     * The provider name.
     *
     * @return The provider name.
     */
    String name();

    /**
     * Prevalence order between each authentication system.
     *
     * @return The order value.
     */
    int order();

    /**
     * Check that the incoming HTTP request can be handle by the underlying authentication system.
     *
     * @param context context data upon which incoming HTTP request can be handled.
     * @return Flag indicating that the current context can be handled by the authentication system.
     */
    boolean canHandle(AuthenticationContext context);

    /**
     * Policies which will be run for each request after authentication method selection
     * The "Security policy chain" may be composed of
     * * {@link PluginAuthenticationPolicy}: a policy based on a plugin
     * * {@link HookAuthenticationPolicy}: a policy based on a class defining the behavior.
     *
     * @return A chain of {@link AuthenticationPolicy}
     */
    List<AuthenticationPolicy> handle(ExecutionContext executionContext);
}
