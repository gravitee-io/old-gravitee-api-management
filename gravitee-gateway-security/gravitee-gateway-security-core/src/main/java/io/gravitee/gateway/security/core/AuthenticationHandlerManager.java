/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.gateway.security.core;


import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/**
 * This class is used to load multiple implementations of {@link AuthenticationHandler}.
 * Loaded implementations are then filtered according to published plan to select only accurate security
 * authentication implementations.
 *
 * @author David BRASSELY (david.brassely at graviteesource.com)
 * @author GraviteeSource Team
 */
public class AuthenticationHandlerManager implements InitializingBean {

    private final Logger logger = LoggerFactory.getLogger(AuthenticationHandlerManager.class);

    @Autowired
    private SecurityProviderLoader securityProviderLoader;

    @Autowired(required = false)
    private AuthenticationHandlerEnhancer authenticationHandlerEnhancer;

    private List<AuthenticationHandler> authenticationHandlers;

    public void afterPropertiesSet() {
        logger.debug("Loading security providers...");
        List<AuthenticationHandler> availableSecurityProviders =
                securityProviderLoader.getSecurityProviders();

        availableSecurityProviders.forEach(authenticationHandler -> {
            if (authenticationHandler instanceof InitializingBean) {
                try {
                    ((InitializingBean) authenticationHandler).afterPropertiesSet();
                } catch (Exception e) {
                    logger.debug("An error occurs while loading security provider [{}]: {}", authenticationHandler.name(), e.getMessage());
                }
            }
        });

        // Filter security providers if a filter is defined
        if (authenticationHandlerEnhancer != null) {
            authenticationHandlers = authenticationHandlerEnhancer.filter(availableSecurityProviders);
        } else {
            authenticationHandlers = availableSecurityProviders;
        }

        // Sort by order
        Collections.sort(authenticationHandlers, Comparator.comparingInt(AuthenticationHandler::order));
    }

    public List<AuthenticationHandler> getAuthenticationHandlers() {
        return authenticationHandlers;
    }

    public void setSecurityProviderLoader(SecurityProviderLoader securityProviderLoader) {
        this.securityProviderLoader = securityProviderLoader;
    }

    public void setAuthenticationHandlerEnhancer(AuthenticationHandlerEnhancer authenticationHandlerEnhancer) {
        this.authenticationHandlerEnhancer = authenticationHandlerEnhancer;
    }
}
