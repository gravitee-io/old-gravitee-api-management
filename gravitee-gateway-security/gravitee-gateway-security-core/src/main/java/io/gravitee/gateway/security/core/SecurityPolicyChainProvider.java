/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.gateway.security.core;

import io.gravitee.common.http.HttpStatusCode;
import io.gravitee.gateway.api.ExecutionContext;
import io.gravitee.gateway.api.buffer.Buffer;
import io.gravitee.gateway.core.processor.StreamableProcessor;
import io.gravitee.gateway.policy.*;
import io.gravitee.gateway.policy.impl.RequestPolicyChain;
import io.gravitee.policy.api.PolicyResult;

import java.util.List;

/**
 * @author David BRASSELY (david.brassely at graviteesource.com)
 * @author GraviteeSource Team
 */
public class SecurityPolicyChainProvider extends AbstractPolicyChainProvider {

    static final String PLAN_UNRESOLVABLE = "GATEWAY_PLAN_UNRESOLVABLE";

    private final PolicyResolver policyResolver;

    public SecurityPolicyChainProvider(final PolicyResolver policyResolver) {
        this.policyResolver = policyResolver;
    }

    @Override
    public StreamableProcessor<ExecutionContext, Buffer> provide(ExecutionContext context) {
        if (context.getAttribute("skip-security-chain") == null) {
            // Calculate the list of policies to apply under this policy chain
            List<Policy> policies = policyResolver.resolve(StreamType.ON_REQUEST, context);

            if (policies == null) {
                return new DirectPolicyChain(
                        PolicyResult.failure(
                                PLAN_UNRESOLVABLE,
                                HttpStatusCode.UNAUTHORIZED_401,
                                "Unauthorized"), context);
            }

            return RequestPolicyChain.create(policies, context);
        }

        return new NoOpPolicyChain(context);
    }
}
