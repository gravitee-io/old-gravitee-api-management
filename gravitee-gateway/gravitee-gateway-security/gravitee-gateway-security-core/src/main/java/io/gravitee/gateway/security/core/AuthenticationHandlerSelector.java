/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.gateway.security.core;

import io.gravitee.gateway.api.Request;

/**
 * @author David BRASSELY (david.brassely at graviteesource.com)
 * @author GraviteeSource Team
 */
@FunctionalInterface
public interface AuthenticationHandlerSelector {

    /**
     * Get an {@link AuthenticationHandler} from the incoming HTTP request.
     * @param request Incoming HTTP request.
     * @return The authentication handler to apply to the incoming request.
     */
    AuthenticationHandler select(Request request);
}
