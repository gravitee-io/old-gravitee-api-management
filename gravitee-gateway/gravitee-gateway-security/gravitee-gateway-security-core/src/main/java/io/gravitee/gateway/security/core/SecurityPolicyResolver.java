/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.gateway.security.core;

import io.gravitee.gateway.api.ExecutionContext;
import io.gravitee.gateway.policy.AbstractPolicyResolver;
import io.gravitee.gateway.policy.Policy;
import io.gravitee.gateway.policy.StreamType;
import org.springframework.beans.factory.annotation.Autowired;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

public class SecurityPolicyResolver extends AbstractPolicyResolver {

    @Autowired
    private AuthenticationHandlerSelector handlerSelector;

    @Override
    public List<Policy> resolve(StreamType streamType, ExecutionContext context) {
        final AuthenticationHandler authenticationHandler = handlerSelector.select(context.request());

        if (authenticationHandler == null) {
            // No authentication method selected, must send a 401
            logger.debug("No authentication handler has been selected to process request {}. Returning an unauthorized status (401)",
                    context.request().id());

            // TODO: it's probably better to throw an exception ?
            return null;
        }

        logger.debug("Authentication handler [{}] has been selected to secure incoming request {}",
                authenticationHandler.name(), context.request().id());

        List<AuthenticationPolicy> policies = authenticationHandler.handle(context);
        return createAuthenticationChain(policies);
    }

    private List<Policy> createAuthenticationChain(List<AuthenticationPolicy> securityPolicies) {
        return securityPolicies.stream().map(new Function<AuthenticationPolicy, Policy>() {
            @Override
            public Policy apply(AuthenticationPolicy securityPolicy) {
                if (securityPolicy instanceof HookAuthenticationPolicy) {
                    try {
                        return (Policy) ((HookAuthenticationPolicy) securityPolicy).clazz().newInstance();
                    } catch (Exception ex) {
                        logger.error("Unexpected error while loading authentication policy", ex);
                    }
                } else if (securityPolicy instanceof PluginAuthenticationPolicy) {
                    return create(
                            StreamType.ON_REQUEST,
                            ((PluginAuthenticationPolicy) securityPolicy).name(),
                            ((PluginAuthenticationPolicy) securityPolicy).configuration());
                }

                return null;
            }
        }).collect(Collectors.toList());
    }

    public void setAuthenticationHandlerSelector(AuthenticationHandlerSelector handlerSelector) {
        this.handlerSelector = handlerSelector;
    }
}
