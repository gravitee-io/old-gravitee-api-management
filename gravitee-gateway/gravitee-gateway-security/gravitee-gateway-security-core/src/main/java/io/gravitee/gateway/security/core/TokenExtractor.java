/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.gateway.security.core;

import io.gravitee.common.http.HttpHeaders;
import io.gravitee.gateway.api.Request;
import org.springframework.util.StringUtils;

import java.util.List;
import java.util.Optional;

/**
 * @author David BRASSELY (david.brassely at graviteesource.com)
 * @author GraviteeSource Team
 */
public class TokenExtractor {

    static final String BEARER = "Bearer";
    static final String ACCESS_TOKEN = "access_token";

    /**
     * Extract JWT from the request.
     * First attempt to extract if from standard Authorization header.
     * If none, then try to extract it from access_token query param.
     * @param request Request
     * @return String Json Web Token.
     */
    public static String extract(Request request) {
        List<String> authorizationHeaders = request.headers() != null ? request.headers().get(HttpHeaders.AUTHORIZATION) : null;

        if (authorizationHeaders != null && !authorizationHeaders.isEmpty()) {
            Optional<String> authorizationBearerHeader = authorizationHeaders
                    .stream()
                    .filter(h -> StringUtils.startsWithIgnoreCase(h, BEARER))
                    .findFirst();
            if (authorizationBearerHeader.isPresent()) {
                String authToken = authorizationBearerHeader.get().substring(BEARER.length()).trim();
                if (!authToken.isEmpty()) {
                    return authToken;
                }
            }
        }

        return request.parameters() != null ? request.parameters().getFirst(ACCESS_TOKEN) : null;
    }
}
