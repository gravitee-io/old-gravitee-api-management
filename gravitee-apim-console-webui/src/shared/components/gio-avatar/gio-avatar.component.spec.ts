/*
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ComponentFixture, TestBed } from '@angular/core/testing';

import { GioAvatarComponent } from './gio-avatar.component';
import { GioAvatarModule } from './gio-avatar.module';

describe('GioAvatarComponent', () => {
  let component: GioAvatarComponent;
  let fixture: ComponentFixture<GioAvatarComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [GioAvatarModule],
    });
    fixture = TestBed.createComponent(GioAvatarComponent);
    component = fixture.componentInstance;
  });

  it('should complete avatar with jdenticon', () => {
    fixture.detectChanges();

    expect(fixture.nativeElement.querySelector('.gio-avatar').innerHTML).toMatchSnapshot();
  });

  it('should display avatar image', () => {
    component.src = 'https://i.pravatar.cc/500';
    fixture.componentInstance.ngAfterViewInit();
    fixture.detectChanges();

    expect(fixture.nativeElement.querySelector('.gio-avatar').innerHTML).toMatchSnapshot();
  });
});
