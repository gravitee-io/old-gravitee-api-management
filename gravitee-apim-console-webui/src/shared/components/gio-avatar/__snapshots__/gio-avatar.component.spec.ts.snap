// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GioAvatarComponent should complete avatar with jdenticon 1`] = `"<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"undefined\\" height=\\"undefined\\" viewBox=\\"0 0 undefined undefined\\" preserveAspectRatio=\\"xMidYMid meet\\"><rect width=\\"100%\\" height=\\"100%\\" fill=\\"#FFFFFF\\" opacity=\\"1.00\\"></rect><path fill=\\"#545454\\" d=\\"M0 0a0,0 0 1,1 0,0a0,0 0 1,1 0,0M0 0a0,0 0 1,1 0,0a0,0 0 1,1 0,0M0 0a0,0 0 1,1 0,0a0,0 0 1,1 0,0M0 0a0,0 0 1,1 0,0a0,0 0 1,1 0,0M0 0a0,0 0 1,1 0,0a0,0 0 1,1 0,0M0 0a0,0 0 1,1 0,0a0,0 0 1,1 0,0M0 0a0,0 0 1,1 0,0a0,0 0 1,1 0,0M0 0a0,0 0 1,1 0,0a0,0 0 1,1 0,0\\"></path><path fill=\\"#d1757a\\" d=\\"M0 0L0 0L0 0L0 0ZM0 0L0 0L0 0L0 0ZM0 0L0 0L0 0L0 0ZM0 0L0 0L0 0L0 0ZM0 0L0 0L0 0L0 0ZM0 0a0,0 0 1,0 0,0a0,0 0 1,0 0,0M0 0L0 0L0 0L0 0ZM0 0a0,0 0 1,0 0,0a0,0 0 1,0 0,0M0 0L0 0L0 0L0 0ZM0 0a0,0 0 1,0 0,0a0,0 0 1,0 0,0M0 0L0 0L0 0L0 0ZM0 0a0,0 0 1,0 0,0a0,0 0 1,0 0,0\\"></path></svg>"`;

exports[`GioAvatarComponent should display avatar image 1`] = `
"<img _ngcontent-a-c11=\\"\\" alt=\\"Avatar picture\\" class=\\"gio-avatar__img\\" src=\\"https://i.pravatar.cc/500\\" width=\\"0\\"><!--bindings={
  \\"ng-reflect-ng-if\\": \\"https://i.pravatar.cc/500\\"
}-->"
`;
